### DISCUSSION TITLE
Create package with API

### DISCUSSION THREAD

**Discussion by georak** [fission/fission#2120] [Q&A]
Hi,

**Question 1**
I am trying to create a package by using API.

below is the request

POST /v2/packages
body:
```json
{
        "metadata": {
            "name": "NAME",
            "namespace": "NAMESPACE"
        },
        "spec": {
            "environment": {
                "namespace": "default",
                "name": "node"
            },
            "deployment": {
                "type": "url",
                "url": "https://raw.githubusercontent.com/fission/fission/master/examples/nodejs/hello.js"
            }
        },
        "status": {
            "buildstatus": "pending"
        }
}
```

The only problem is that package state is "running" and after 5 minutes it fails with this error:
max retries exceeded in building source package, timeout due to environment builder not ready

pkg info

![image](https://user-images.githubusercontent.com/3660446/126498148-8c14416f-e6b1-4137-9c83-e88b5cb1034d.png)

Do I have to change something on the payload to make it work?

**Question 2**
Can I pass the code as string like this?

 "deployment": {
                "type": "literal",
                "url": CODE_AS_BASE64_ENCODED_STRING
            }

**Comment by sanketsudake** [comment]
@georak  If you are using type literal you need to pass literal value in the payload.
You can refer to CRD spec here. https://doc.crds.dev/github.com/fission/fission/fission.io/Package/v1@v1.13.1

You can try  `fission pkg info --name <pkg_name>` to more info about package and its build logs.

Also, there is a rebuild command if it helps.
`fission pkg rebuild`

**Reply by georak** [reply]
@sanketsudake thank you for the response.
Is the payload that I send correct?

I have tried everything but still package is everytime on "running" state and after 5 minutes it fails. Rebuild gie exactly the same result. 

Both literal and url types have the same result, so I think that I have to pass something more on the POST payload.

Below you can see the info

![image](https://user-images.githubusercontent.com/3660446/126601636-df6d354d-5da7-4237-b6c3-d4cbf7218a38.png)

**Reply by georak** [reply]
Also I am using v2, do you have the CRD spec for v2?

**Reply by georak** [reply]
I found it!! As I am passing deployment and not source, I had to pass buildStatus as 'succeded' everything is working. Now I have a python API that can create functions :)

**Reply by sanketsudake** [reply]
Nice 👍  We dont have v2 CRDs, only v1. 

https://doc.crds.dev/github.com/fission/fission
